### PR TITLE
Add top n collection and refactor raw query logic

### PR DESCRIPTION
**Pull request by xichen2020**
cc  @black-adder @notbdu

This PR adds the generic `topNValues` collection type to keep the top N values as it is needed for both raw query and grouped query processing. Also refactored how the raw results store values internally and changed it to use a `topNValues` collection instead of always keeping values sorted to make it more efficient and the code structure cleaner.

### CODE REVIEW

File: query/raw_result.go [hunk 1]
```diff
@@ -131,121 +74,94 @@ func (r *RawResults) IsComplete() bool { return r.LimitReached() && !r.IsOrdered
...
 // MaxOrderByValues returns the orderBy field values for the largest result in
 // the result collection.
-func (r *RawResults) MaxOrderByValues() []field.ValueUnion {
+func (r *RawResults) MaxOrderByValues() field.Values {
 	if r.Len() == 0 {
 		return nil
 	}
-	return r.Data[r.Len()-1].OrderByValues
+	// NB: The "min" element in `Ordered` is actually the largest element among the top N results
+	// because the heap uses `ResultRverseLessThanFn` for ordering purposes.
```
Comment (black-adder) [inline query/raw_result.go]: ResultR**e**verseLessThanFn

Comment (black-adder) [reply query/raw_result.go]: Also, this whole thing is kinda confusing (min being the max element), maybe rename `Min()` to `Top()`? `Min()` feels misleading while `Top()` can semantically mean anything.

File: query/raw_result.go [hunk 2]
```diff
@@ -7,109 +7,52 @@ import (
...
-	return it.currIdx < len(it.resultsByDocIDAsc)
-}
-
-// DocID returns the doc ID of the current raw result.
-func (it *RawResultIterator) DocID() int32 { return it.resultsByDocIDAsc[it.currIdx].DocID }
-
-// Values returns the collection of values to order the current raw results by.
-func (it *RawResultIterator) Values() []field.ValueUnion {
-	return it.resultsByDocIDAsc[it.currIdx].OrderByValues
+	Data          string
```
Comment (black-adder) [inline query/raw_result.go]: what is data here? the raw source field? If so, can we rename it or leave a comment?

File: storage/segment_query.go [hunk 3]
```diff
@@ -215,9 +215,15 @@ func collectOrderedRawResults(
 	}
 	defer filteredOrderByIter.Close()
 
+	// NB: We currently first determine the doc ID set eligible based on orderBy criteria
+	// and then fetch the corresponding raw doc source data, instead of iterating over
+	// both the orderBy fields and the raw doc source field in sequence. This is because
+	// the raw doc source field is typically much larger in dataset size and expensive to
+	// decode. As such we restrict the amount of decoding we need to do by restricting the
+	// doc ID set first.
 	orderedRawResults, err := collectTopNRawResultDocIDOrderByValues(
```
Comment (black-adder) [inline storage/segment_query.go]: thanks for the comment, demystifies things

File: storage/segment_query.go [hunk 4]
```diff
@@ -269,112 +275,71 @@ func createFilteredOrderByIterator(
...
 // collectTopNRawResults collects the top N raw results from the raw doc source field
-// based on the doc IDs and the ordering specified in `orderdRawResults`. The result
-// array returned contains raw results ordered in the same order as that dictated by the
-// `orderBy` clauses (e.g., if `orderBy` requires results by sorted by timestamp in descending order,
-// the result array will also be sorted by timestamp in descending order).
+// based on the doc IDs and the ordering specified in `res`.
 func collectTopNRawResults(
 	rawDocSourceField indexfield.StringField,
-	orderedRawResults []query.RawResult,
+	topNDocIDValues []docIDValues, // Owned but not sorted in any order
```
Comment (black-adder) [inline storage/segment_query.go]: period?

File: storage/segment_query.go [hunk 5]
```diff
@@ -269,112 +275,71 @@ func createFilteredOrderByIterator(
...
 	if err != nil {
 		return fmt.Errorf("error fetching raw doc source data: %v", err)
 	}
-	numItemsWithRawDocSource := 0
 	for rawDocSourceIter.Next() {
-		maskingPos := rawDocSourceIter.MaskingPosition()
-		orderedRawResults[maskingPos].HasData = true
-		orderedRawResults[maskingPos].Data = rawDocSourceIter.Value()
-		numItemsWithRawDocSource++
+		maskingPos := rawDocSourceIter.MaskingPosition() // Position in the (doc ID, values) sequence
```
Comment (black-adder) [inline storage/segment_query.go]: ditto

File: storage/segment_query.go [hunk 6]
```diff
@@ -637,3 +620,74 @@ func closeIteratorsOnError(iters []indexfield.BaseFieldIterator, err error) erro
...
+}
+
+// cloneDocIDValues clones the incoming (doc ID, values) pair.
+func cloneDocIDValues(v docIDValues) docIDValues {
+	// TODO(xichen): Should pool and reuse the value array here.
+	v.Values = v.Values.Clone()
+	return v
+}
+
+// cloneDocIDValuesTo clonse the (doc ID, values) pair from `src` to `target`.
```
Comment (black-adder) [inline storage/segment_query.go]: clonse

File: query/raw_result.go [hunk 7]
```diff
@@ -7,109 +7,52 @@ import (
...
+	OrderByValues field.Values // For ordering purposes, Empty for unsorted raw results
```
[the remark this answers is not in the document]
Comment (xichen2020) [reply query/raw_result.go]: So the convention is to end comments w/ periods for comments above functions and type defs, but not end them w/ periods for inline comments on the same line as the parameter definitions.

[the remark this answers is not in the document]
Comment (xichen2020) [reply query/raw_result.go]: Will fix the `E`.

File: query/raw_result.go (hunk 1 above)
Comment (xichen2020) [reply query/raw_result.go]: Sounds good can do.

File: query/raw_result.go (hunk 2 above)
Comment (xichen2020) [reply query/raw_result.go]: Yeah can leave a comment.

File: storage/segment_query.go (hunk 4 above)
Comment (xichen2020) [reply storage/segment_query.go]: See comments above

File: storage/segment_query.go (hunk 6 above)
Comment (xichen2020) [reply storage/segment_query.go]: Good catch, will fix.

File: query/raw_result.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (black-adder) [reply query/raw_result.go]: can you link me to the definitive documentation for golang documentation? I'd rather not attempt to be a smartass and fail again.

[the remark this answers is not in the document]
Comment (xichen2020) [reply query/raw_result.go]: Hey so there's no definitive doc on this, I'm mostly just following the style in the go src code (which is itself inconsistent sometimes). Example: https://golang.org/src/io/pipe.go

**Pull request merged**